### ISSUE TITLE
Generic IMapChange

### ISSUE THREAD

**Issue by hediet** [mobxjs/mobx#847]
Even though IMapChange is generic, it does not really use its type parameter:
```ts
export interface IMapChange<T> {
	object: ObservableMap<T>;
	type: "update" | "add" | "delete";
	name: string;
	newValue?: any;
	oldValue?: any;
}
```
(https://github.com/mobxjs/mobx/blob/master/src/types/observablemap.ts#L44)

Is there a reason why `newValue` and `oldValue` aren't of type T?

BTW.: I really like mobx and I couldn't imagine how to get along without it! Thanks for your effort!

**Comment by mweststrate** [comment]
@hediet nope, I think that is just a missing type! feel free to submit a PR

**Comment by mweststrate** [comment]
Note that the type would be ` T | undefined`, or to nail it precisely down, the interface could be splitted up further (for `"add"` oldValue is always `undefined` and newValue is `T` and never `undefined`, etc)

**Comment by tcrognon** [comment]
How about something like this?

```ts
type IMapChange<K, T extends keyof IMapChangeTypeValueLookup<K>> = {
    object: ObservableMap<K>;
    type: T;
    name: string;
} & IMapChangeTypeValueLookup<K>[T];

interface IMapChangeTypeValueLookup<K> {
    update: {
        oldValue: K;
        newValue: K;
    };
    add: {
        newValue: K;
    };
    delete: {
        oldValue: K;
    };
}
```

**Comment by hediet** [comment]
Yeah, `keyof` rules :D 
However, `keyof` does not play very well with type inference:
```typescript
const c = null as IMapChange<any, "delete" | "add" | "update">;
if (c.type === "add") {
    c. // does not suggest newValue
}
```

This is why I would suggest to make it more explicit:
```typescript
export type IMapChange<T> = IMapChangeUpdate<T> | IMapChangeAdd<T> | IMapChangeDelete<T>;

interface IMapChangeBase<T> {
  object: ObservableMap<T>;
  name: string;
}

export interface IMapChangeUpdate<T> extends IMapChangeBase<T> {
	type: "update";
	newValue: T;
	oldValue: T;
}

export interface IMapChangeAdd<T> extends IMapChangeBase<T> {
	type: "add";
	newValue: T;
}

export interface IMapChangeDelete<T>  extends IMapChangeBase<T> {
	type: "delete";
	oldValue: T;
}
```

**Comment by mweststrate** [comment]
@hediet yep I think that is the way to go. Interested in submitting a PR?

**Comment by hediet** [comment]
Yeah, I will do that today. Sorry for the delay.

**Comment by mweststrate** [comment]
Released as ` 3.1.5`. Thanks for the PR!
